src/utp_stream.cpp

```diff
@@ -2850,7 +2839,9 @@ bool utp_socket_impl::incoming_packet(boost::uint8_t const* buf, int size
 
 	// if the socket is closing, always ignore any packet
 	// with a higher sequence number than the FIN sequence number
-	if (m_eof && compare_less_wrap(m_eof_seq_nr, ph->seq_nr, ACK_MASK))
+	// ST_STATE messages always include the next seqnr.
+	if (m_eof && (compare_less_wrap(m_eof_seq_nr, ph->seq_nr, ACK_MASK)
+		|| (m_eof_seq_nr == ph->seq_nr && ph->get_type() != ST_STATE)))
```

Comment:
it's a bit awkward, but `ST_STATE` messages use the "next" sequence number. So, if the packet you receive is `ST_STATE`, the sequence number indicates what will be used in the next `ST_DATA` message. So, when shutting down, it's OK to receive more messages with the same sequence number as the `FIN`, as long as they are `ST_STATE` messages